File: fpga/openxc7-synth/riscv/test_program.zig
// ╔════════════════════════════════════════════════════════════════════════════╗
// ║  TRINITY OS — RISC-V Test Program (Zig)                                     ║
// ║                                                                              ║
// ║  Simple Zig program to verify RISC-V CPU on FPGA                            ║
// ║                                                                              ║
// ║  Build: zig build-exe test_program.zig \                                   ║
// ║           -target riscv32-none-none -mcpu=generic_rv32 \                    ║
// ║           -fno-builtin -fno-stdlib -O ReleaseFast                           ║
// ╚════════════════════════════════════════════════════════════════════════════╝

const std = @import("std");

// Memory-mapped I/O addresses
const LED_CTRL: *volatile u32 = @ptrFromInt(0x1000);
const UART_TX: *volatile u32 = @ptrFromInt(0x2000);
const UART_STATUS: *volatile u32 = @ptrFromInt(0x2004);

// Golden identity: φ² + 1/φ² = 3
const PHI: f64 = 1.6180339887498948482;
const GOLDEN_IDENTITY: f64 = PHI * PHI + 1.0 / (PHI * PHI);

// Simple delay loop
inline fn delay(count: u32) void {
    var i: u32 = 0;
    while (i < count) : (i += 1) {
        asm volatile ("nop");
    }
}

// Fibonacci calculation (optimized tail recursion)
fn fibonacci(n: u32) u32 {
    if (n <= 1) return n;

    var a: u32 = 0;
    var b: u32 = 1;
    var i: u32 = 2;

    while (i <= n) : (i += 1) {
        const c = a + b;
        a = b;
        b = c;
    }

    return b;
}

// Test memory read/write
fn testMemory() bool {
    const test_addr: *volatile u32 = @ptrFromInt(0x3000);

    // Write and read back
    test_addr.* = 0xDEADBEEF;
    test_addr.* = 0x12345678;

    const val = test_addr.*;
    return val == 0x12345678;
}

// Lucas number (TRINITY sequence)
// Lucas: 2, 1, 3, 4, 7, 11, 18, 29, 47, 76, 123...
// Note: L(2) = 3 = TRINITY
fn lucas(n: u32) u32 {
    if (n == 0) return 2;
    if (n == 1) return 1;

    var a: u32 = 2;
    var b: u32 = 1;
    var i: u32 = 2;

    while (i <= n) : (i += 1) {
        const c = a + b;
        a = b;
        b = c;
    }

    return b;
}

// Blink LED pattern
fn blinkPattern(count: u32, fast: bool) void {
    const delay_val: u32 = if (fast) 20000 else 100000;

    var i: u32 = 0;
    while (i < count) : (i += 1) {
        LED_CTRL.* = 1;
        delay(delay_val);
        LED_CTRL.* = 0;
        delay(delay_val);
    }
}

// Main entry point
export fn _start() noreturn {
    // Test calculations
    const fib_result = fibonacci(10);  // 55
    const lucas_result = lucas(10);    // 123
    const mem_ok = testMemory();

    // Infinite loop with TRINITY sacred pattern
    var counter: u32 = 0;

    while (true) {
        counter +%= 1;

        // Pattern 1: 3 blinks (TRINITY)
        blinkPattern(3, false);

        // Pattern 2: Fibonacci mod 10 blinks
        const fib_count = @as(u32, @intCast(fib_result % 10));
        blinkPattern(fib_count, true);

        // Pattern 3: Lucas mod 7 blinks (sacred numbers)
        const lucas_count = @as(u32, @intCast(lucas_result % 7));
        blinkPattern(lucas_count, true);

        // Pause
        delay(500000);

        // Verify golden identity every cycle
        comptime {
            if (GOLDEN_IDENTITY < 2.99 or GOLDEN_IDENTITY > 3.01) {
                @compileError("Golden Identity violated!");
            }
        }
    }
}

// ╔════════════════════════════════════════════════════════════════════════════╗
// ║  EXPECTED LED PATTERN                                                        ║
// ╠════════════════════════════════════════════════════════════════════════════╣
// ║  1. Three slow blinks (TRINITY = 3)                                         ║
// ║  2. Five fast blinks (Fibonacci 10 = 55 → 55 % 10 = 5)                     ║
// ║  3. Four fast blinks (Lucas 10 = 123 → 123 % 7 = 4)                         ║
// ║  4. Pause                                                                   ║
// ║  5. Repeat forever                                                          ║
// ╚════════════════════════════════════════════════════════════════════════════╝

File: fpga/openxc7-synth/riscv/test_program.c
// ╔════════════════════════════════════════════════════════════════════════════╗
// ║  TRINITY OS — RISC-V Test Program                                          ║
// ║                                                                              ║
// ║  Simple C program to verify RISC-V CPU is working on FPGA                   ║
// ║  Features:                                                                  ║
// ║  - Fibonacci calculation (tests ALU, branches)                             ║
// ║  - Memory read/write (tests data bus)                                      ║
// ║  - Infinite loop with pattern (tests instructions fetch)                   ║
// ║                                                                              ║
// ║  Compile: riscv32-unknown-elf-gcc -march=rv32imc -mabi=ilp32 -nostdlib    ║
// ║          -Wl,-Ttext=0x0 -o test_program.elf test_program.c                 ║
// ║  Extract: riscv32-unknown-elf-objcopy -O binary test_program.elf prog.bin  ║
// ╚════════════════════════════════════════════════════════════════════════════╝

// Volatile pointer to LED output (memory-mapped I/O)
// Address 0x1000 = LED control register
volatile unsigned int* const LED_CTRL = (unsigned int*)0x1000;

// Simple delay loop
void delay(volatile unsigned int count) {
    while (count--) {
        __asm__ volatile ("nop");
    }
}

// Fibonacci calculation (iterative)
unsigned int fibonacci(unsigned int n) {
    if (n <= 1)
        return n;

    unsigned int a = 0, b = 1, c, i;
    for (i = 2; i <= n; i++) {
        c = a + b;
        a = b;
        b = c;
    }
    return b;
}

// Test memory read/write
unsigned int test_memory(void) {
    volatile unsigned int* test_addr = (unsigned int*)0x2000;

    // Write pattern
    *test_addr = 0xDEADBEEF;
    *test_addr = 0x12345678;

    // Read back
    unsigned int val = *test_addr;

    return (val == 0x12345678) ? 1 : 0;
}

// Main function
void _start(void) {
    unsigned int counter = 0;
    unsigned int fib_result;
    unsigned int mem_ok;

    // Test 1: Fibonacci (ALU + branches)
    fib_result = fibonacci(10);  // Should be 55

    // Test 2: Memory (data bus)
    mem_ok = test_memory();

    // Main loop: LED blink pattern
    while (1) {
        // Pattern: Fibonacci sequence on LED
        counter++;

        // Blink pattern
        for (int i = 0; i < 5; i++) {
            *LED_CTRL = 0x01;  // LED on
            delay(100000);
            *LED_CTRL = 0x00;  // LED off
            delay(100000);
        }

        // Fast blink (fibonacci count)
        for (int i = 0; i < (fib_result % 10); i++) {
            *LED_CTRL = 0x01;
            delay(20000);
            *LED_CTRL = 0x00;
            delay(20000);
        }

        // Longer pause
        delay(500000);
    }
}

// ╔════════════════════════════════════════════════════════════════════════════╗
// ║  EXPECTED BEHAVIOR                                                          ║
// ╠════════════════════════════════════════════════════════════════════════════╣
// ║  1. 5 normal blinks                                                          ║
// ║  2. Fibonacci(10) = 55 → 5 fast blinks                                      ║
// ║  3. Pause                                                                    ║
// ║  4. Repeat forever                                                           ║
// ╚════════════════════════════════════════════════════════════════════════════╝
